test/framework/exec/docker.go

```diff
@@ -0,0 +1,360 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+var dockerClient *client.Client
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	if dockerClient == nil {
+		var err error
+		dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return dockerClient, nil
+}
+
+// SaveImage saves a Docker image to the file specified by dest.
+func SaveDockerImage(ctx context.Context, image, dest string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	reader, err := docker.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
+
+	_, err = io.Copy(tar, reader)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PullDockerImage triggers the Docker engine to pull an image if it is not
+// already present.
+func PullDockerImage(ctx context.Context, image string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	pullResp, err := docker.ImagePull(ctx, image, types.ImagePullOptions{})
+	if err != nil {
+		return err
+	}
+	defer pullResp.Close()
+
+	// Clients must read the ImagePull response to EOF to complete the pull
+	// operation or errors can occur.
+	_, err = io.ReadAll(pullResp)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// GetLoadBalancerHostPort looks up the host port bound for the load balanced API.
+func GetLoadBalancerHostPort(ctx context.Context, name string) (string, error) {
+	docker, err := getDockerClient()
+	if err != nil {
+		return "", err
+	}
+
+	// Get details about the container
+	containerInfo, err := docker.ContainerInspect(ctx, name)
+	if err != nil {
+		return "", err
+	}
+
+	// Loop through the container port bindings and return the first HostPort
+	for port, bindings := range containerInfo.NetworkSettings.Ports {
+		if port == "6443/tcp" {
+			for _, binding := range bindings {
+				return binding.HostPort, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("no host port found for load balancer '%s'", name)
+}
+
+// ContainerExecToFile executes a command in a running container and writes any output to the provided fileOnHost.
+func ContainerExecToFile(ctx context.Context, containerName string, fileOnHost *os.File, command string, args ...string) error {
```

Comment:
I think we should decide which way we go with the func names. Probably choosing either Container or Docker is better. (This also depends on if we want to change the package name)

Reply:
Good thing to think about, especially if we ever want to expand this to work with more than just Docker. At least for the time being, I've renamed `SaveDockerImage` and `PullDockerImage` to `SaveContainerImage` and `PullContainerImage` so the function calls are consistently referring to just "container".